tools/codegen/api/python.py

```diff
@@ -0,0 +1,987 @@
+    #
+    # TODO: this seems too complicated - maybe we can simplify after full
+    # c10 dispatch migration?
+    typename = la.cpp_type_str \
+        if a.name != 'out' and la is not None else a.cpp_type_str
+
+    if a.default_init is not None:
+        # Note: only introduced in tensor_options_args
+        if typename not in UNPACK_WITH_DEFAULT_METHODS:
+            raise RuntimeError(
+                f'type \'{typename}\' is not supported in default_init')
+        unpack_with_default = UNPACK_WITH_DEFAULT_METHODS[typename]
+        expr = f'_r.{unpack_with_default}({arg_index}, {a.default_init})'
+    elif a.size is not None:
+        if typename not in UNPACK_WITH_SIZE_METHODS:
+            raise RuntimeError(
+                f'type \'{typename}\' with definite size ({a.size}) is not supported')
+        unpack_with_size = UNPACK_WITH_SIZE_METHODS[typename].format(a.size)
+        expr = f'_r.{unpack_with_size}({arg_index})'
+    else:
+        unpack = UNPACK_METHODS.get(typename)
+        if unpack is None:
+            raise RuntimeError(f'type \'{typename}\' is not supported')
+        expr = f'_r.{unpack}({arg_index})'
+
+    return PythonArgParserOutputExpr(
+        name=a.name,
+        expr=expr,
+        index=arg_index,
+        argument=a,
+    )
+
+# Returns a map with key = arg_name and value = PythonArgParserOutputExpr.
+def arg_parser_output_exprs(
+    ps: PythonSignature, f: NativeFunction, *, method: bool
+) -> Dict[str, PythonArgParserOutputExpr]:
+    lambda_args = dispatch_lambda_args(f, method, python_signature=ps)
+    lambda_args_map = dict(map(lambda a: (a.name, a), lambda_args))
+
+    return dict(map(lambda e: (e.name, e),
+                    map(lambda i: arg_parser_output_expr(i[0], i[1], lambda_args_map.get(i[1].name)),
+                        enumerate(ps.arguments()))))
+
+# argument name to 'simple_type' for scattered tensor options fields
+TENSOR_OPTIONS_FIELDS = {
+    'dtype': 'ScalarType',
+    'device': 'Device',
+    'layout': 'Layout',
+    'pin_memory': 'bool',
+    'requires_grad': 'bool',
+}
+
+# bind arg parser outputs (python args) with dispatch lambda arguments (c++ args).
+def dispatch_lambda_exprs(
+    ps: PythonSignature, f: NativeFunction, *, method: bool
+) -> Tuple[
+    LocalInitExprs,
+    DispatchLambdaArgumentExprs,
+]:
+    # This method is to bind 'arg_parser_outputs' and 'lambda_args' by producing
+    # 'inits' and 'lambda_args_exprs' for each lambda argument using arg parser
+    # outputs.
+    arg_parser_outputs = arg_parser_output_exprs(ps, f, method=method)
+    lambda_args = dispatch_lambda_args(f, method, python_signature=ps)
+    inits: List[str] = []
+    lambda_args_exprs: Dict[str, str] = dict()
+
+    has_toptions = has_tensor_options(f)
+
+    # 1. special inits/unpacking to provide binding exprs for lambda arguments.
+    for a in ps.arguments(skip_tensor_options=True):
+        name = a.name
+        arg_parser_expr = arg_parser_outputs[a.name].expr
```

Comment:
Not sure what action to take here, but it is a little weird we have to indirect through a dictionary to find these expressions.